pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go

```diff
@@ -0,0 +1,993 @@
+	default:
+		return nil, fmt.Errorf("unsupported load balancer affinity: %v", affinity)
+	}
+
+	sourceRanges, err := service.GetLoadBalancerSourceRanges(apiService)
+	if err != nil {
+		return nil, err
+	}
+
+	if !service.IsAllowAll(sourceRanges) {
+		return nil, fmt.Errorf("Source range restrictions are not supported for openstack load balancers")
+	}
+
+	glog.V(2).Infof("Checking if openstack load balancer already exists: %s", cloudprovider.GetLoadBalancerName(apiService))
+	_, exists, err := lbaas.GetLoadBalancer(apiService)
+	if err != nil {
+		return nil, fmt.Errorf("error checking if openstack load balancer already exists: %v", err)
+	}
+
+	// TODO: Implement a more efficient update strategy for common changes than delete & create
+	// In particular, if we implement hosts update, we can get rid of UpdateHosts
+	if exists {
+		err := lbaas.EnsureLoadBalancerDeleted(apiService)
+		if err != nil {
+			return nil, fmt.Errorf("error deleting existing openstack load balancer: %v", err)
+		}
+	}
+
+	lbmethod := v2_pools.LBMethod(lbaas.opts.LBMethod)
+	if lbmethod == "" {
+		lbmethod = v2_pools.LBMethodRoundRobin
+	}
+	name := cloudprovider.GetLoadBalancerName(apiService)
+
+	createOpts := loadbalancers.CreateOpts{
+		Name:        name,
+		Description: fmt.Sprintf("Kubernetes external service %s", name),
+		VipSubnetID: lbaas.opts.SubnetId,
+	}
+
+	loadBalancerIP := apiService.Spec.LoadBalancerIP
+	if loadBalancerIP != "" {
+		createOpts.VipAddress = loadBalancerIP
+	}
+
+	loadbalancer, err := loadbalancers.Create(lbaas.network, createOpts).Extract()
+	if err != nil {
+		// cleanup what was created so far
+		_ = lbaas.EnsureLoadBalancerDeleted(apiService)
+		return nil, err
+	}
+
+	waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
+
+	listener, err := listeners.Create(lbaas.network, listeners.CreateOpts{
+		Name:           name,
+		Protocol:       listeners.ProtocolTCP,
+		ProtocolPort:   (int)(ports[0].Port), //TODO: need to handle multi-port
+		LoadbalancerID: loadbalancer.ID,
+	}).Extract()
+	if err != nil {
+		// cleanup what was created so far
+		_ = lbaas.EnsureLoadBalancerDeleted(apiService)
+		return nil, err
+	}
+
+	waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
+
+	pool, err := v2_pools.Create(lbaas.network, v2_pools.CreateOpts{
+		Name:        name,
+		Protocol:    v2_pools.ProtocolTCP,
+		LBMethod:    lbmethod,
+		ListenerID:  listener.ID,
+		Persistence: persistence,
+	}).Extract()
+	if err != nil {
+		// cleanup what was created so far
+		_ = lbaas.EnsureLoadBalancerDeleted(apiService)
+		return nil, err
+	}
+
+	for _, host := range hosts {
+		addr, err := getAddressByName(lbaas.compute, host)
+		if err != nil {
+			return nil, err
+		}
+
+		waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
+
+		_, err = v2_pools.CreateAssociateMember(lbaas.network, pool.ID, v2_pools.MemberCreateOpts{
+			Name:         name,
```

Comment:
there is no `name` for pool members according to https://wiki.openstack.org/wiki/Neutron/LBaaS/API_2.0#Add_a_New_Member_to_a_Pool, I got the error "create failed (client error): Unrecognized attribute(s) 'name'" when creating a service.

I had to remove this line to be able to successfully create a service with lbv2 on openstack liberty